---
title: 基础巩固第二天
date: 2021-12-16T03:47:59.944Z
---
1. 异常体系

   1. 发现错误的最佳时机是编译时，编译期找不到的叫运行时异常`RuntimeException`

      ![](images/异常.png)
   2. 打印异常信息

      ```java
      catch(Exception e){
          e.printStackTrace();                         //全部错误信息,可以输出到流(PrintStream s)
          System.out.println(e.getMessage());          //错误返回内容
          System.out.println(e.getLocalizedMessage()); //错误返回内容
          System.out.println(e.toString());            //异常类：返回内容
          System.out.println(e.fillInStackTrace());    //异常类：返回内容
      }
      ```
   3. RuntimeException

      1. 数组越界异常

         ```java
         try{
             int[] i = {1,2,3,4};
             System.out.println(i[4]);
         }catch(ArrayIndexOutOfBoundsException e){
             System.out.println("数组越界");
         }
         ```
      2. 空指针异常

         ```java
         try{
           Two two = new Two();
           two=null;
           System.out.println(two.id);
         }catch(NullPointerException e){
           System.out.println("空指针异常");
         }
         ```
      3. 数组长度为负异常

         ```java
         try{
           int [] j = new int[-1];
         }catch(NegativeArraySizeException e){
           System.out.println("数组长度为负异常");
         }
         ```
      4. 类型转换异常

         ```
         class Father{
           class Son1 extends Father{
           }
           class Son2 extends Father{
           }
           
           try{
             Father son1 = new Son1();
             Son2 son2 (Son2)son1;
           }catch(ClassCastException e){
             System.out.println("类型转换异常");
           }
         }
         ```